langserver/langserver.py

```diff
@@ -358,6 +363,22 @@ def serve_x_definition(self, request):
                     "file": filename
                 }
 
+            else:
+                # defining_module cannot be found anywhere in the workspace, but it might
+                # be elsewhere in the filesystem
+                filename = os.path.basename(defining_module_path)
+                symbol_name, symbol_kind = "", ""
+                if d.description:
+                    symbol_name, symbol_kind = LangServer.name_and_kind(d.description)
+                symbol_locator["symbol"] = {
+                    "name": symbol_name,
+                    "kind": symbol_kind,
+                    "file": filename,
+                }
+                package = get_pip_package(defining_module_path)
+                if package is not None:
+                    symbol_locator["symbol"]["package"] = { "name": package.key }
```

Comment:
should this logic live in `Workspace`?

Reply:
I'm not sure. Currently, it looks like the `Workspace` instance only worries about files within its file tree or dependencies that it manages. This refers to files elsewhere in the filesystem (likely dependencies manually installed by the user). It's unclear to me whether the logic to infer the package of a symbol defined elsewhere in the filesystem should fall to the `Workspace`. Whatever we decide, it would be nice to add a docstring for the `Workspace` class after this to describe its responsibilities.

Maybe @akhleung can weigh in?

Reply:
Hmm ... I would say that if those external files eventually need to be traversed and indexed in order to make code intelligence work, then it's fine to put it in `Workspace`, since I see that class's responsibility as keeping track of all dependencies, package names, etc.

Reply:
Just to clarify I'm thinking about this correctly. The proposal would be to eliminate the `else` clause here and add handling to `Workspace.get_module_by_path` to deal with external modules. This method should be modified so that, in the case the module is not found locally, it is found via the logic that used to be in this `else` clause.

If that's the case, a few follow-up questions:
- What are the semantics of the fields in the `Module` class (defined in `workspace.py`)? And in particular, what does `is_external` indicate and would that be applicable in this case?